### PR TITLE
Verifyta

### PR DESCRIPTION
**Pull request by frejaz**
## Description
The complete implementation of the verifiers microservice.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Feature that assists further development (non-breaking change for the developer)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

### How Has This Been 
Dynamic white and black box tests.
Monkey tests.

### Checklist:

- [X] You have added a clear title and description for the PR
- [X] The CI pass
- [X] There are no merge conflicts
- [x] The code satisfies the [code standards](https://docs.google.com/document/d/1R9n9MeuRwz3WSrnWZUjrbmGeosTlu1Xz8f5F7gu2AJE/edit#heading=h.e9xl4xfoldow) 
- [X] I have added docstrings for my code, where needed
- [X] Code implements intended functionality
- [X] All relevant tests pass (Black, White, Integration, Monkey, etc.)

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -9,7 +9,8 @@
   ],
   "rules": {
     "no-console": 1,
-    "prettier/prettier": 2
+    "prettier/prettier": 1,
+    "endOfLine": 0 //0 = off, 1 = warn, 2 = error
```
Comment (Brandhoej) [inline .eslintrc:13]: Hvorfor er det ændret?

File: build/index.js [hunk 2]
```diff
@@ -1,3 +1,66 @@
 "use strict";
```
Comment (Brandhoej) [inline build/index.js]: /build burde være i .gitignore og .dockerignore

File: package.json [hunk 3]
```diff
@@ -49,7 +49,14 @@
...
+    "typedoc": "^0.23.20",
     "typescript": "^4.7.4"
   },
-  "dependencies": {}
+  "dependencies": {
+    "crypto-js": "^4.1.1",
+    "dotenv": "^16.0.3",
+    "fastify": "^4.9.2",
+    "fastify-xml-body-parser": "^2.1.1",
+    "memfs": "^3.4.9"
```
Comment (Brandhoej) [inline package.json:60]: Bliver de her dependencies brugt?

File: src/infrastructure/currentRequestXml/empty.js [hunk 4]
```diff
@@ -0,0 +1 @@
+// this is empty file, so that this directory is added to build
```
Comment (Brandhoej) [inline src/infrastructure/currentRequestXml/empty.js:1]: Kan man ikke bare få den til at lave det directory hvis det mangler?

File: src/infrastructure/i_cmd_result.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+export interface ICmdResult {
+  verifierOutput: string; // If there is an XML error, this will be an empty string
```
Comment (Brandhoej) [inline src/infrastructure/i_cmd_result.ts:2]: Kommentaren burde være en docstring

File: src/infrastructure/i_cmd_result.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+  verifierError: string;
+  cmdError?: string;
```
Comment (Brandhoej) [inline src/infrastructure/i_cmd_result.ts:4]: Hvor kan ``verifierError`` ikke være undefined, men ``cmdError`` kan?

File: src/infrastructure/paths.spec.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
+import {
+  rootPath,
+  verifytaLinuxPath,
+  verifytaTestModelsPath,
+  verifytaWindowsPath,
+} from './paths';
+
+import Os from 'os';
+const SLASH = Os.platform() === 'linux' ? '/' : '\\';
```
Comment (Brandhoej) [inline src/infrastructure/paths.spec.ts]: Den bliver ikke brugt?

File: src/infrastructure/paths.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
+import Os from 'os';
+import path from 'path';
+
+const SLASH = Os.platform() === 'linux' ? '/' : '\\';
```
Comment (Brandhoej) [inline src/infrastructure/paths.ts:4]: Den bliver ikke brugt?

File: src/infrastructure/verifyta_environment.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        return res;
+      })
+      .catch((error) => {
+        Promise.reject(error);
+      }) as Promise<ICmdResult>;
+
+    return result;
+  }
+
+  // TODO: input string is not formatted correctly, so file is saved formatting badly...
```
Comment (Brandhoej) [inline src/infrastructure/verifyta_environment.ts:44]: TODO?

File: src/infrastructure/verifyta_environment.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+  tempSaveFile(contents: string): string {
+    // Create unique name for the xmlfile
+    const hashedFilename =
+      crypto
+        .createHash('md5')
+        .update(contents + Math.random())
+        .digest('hex') + '.xml';
+
+    // Save temerary file
+    const seperator = Os.platform() === 'linux' ? '/' : '\\';
```
Comment (Brandhoej) [inline src/infrastructure/verifyta_environment.ts:62]: Jeg har set det her ret mange steder.

File: src/infrastructure/verifyta_output_parser.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+export class VerifytaOutputParser {
+  /**
+   * Parses the ICmdResult created by the VerifytaEnvironment.execute() function.
+   * Will create an IQueryResult based on the queries parsing or failing and whether
+   * there are syntax errors or parser errors.
+   * @param verifytaOuput the output from VerifytaEnvironments execute
+   * @param queries An array of the names of each query run
+   * @returns the parsed result as an IQueryResult
+   */
+  parse(verifytaOuput: ICmdResult, queries: Array<string>): IQueryResult {
```
Comment (Brandhoej) [inline src/infrastructure/verifyta_output_parser.ts:14]: Shorthand for array "[]" er bedre.

File: src/infrastructure/verifyta_verifier.ts [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+  }
+
+  async verifySolution(
+    xmlFile: string,
+    queries: Array<string>,
+  ): Promise<IQueryResult> {
+    const result = await this.environment.execute(xmlFile);
+    const parsedResult = this.parser.parse(result, queries);
+
+    return parsedResult;
```
Comment (Brandhoej) [inline src/infrastructure/verifyta_verifier.ts:34]: ``return this.parser.parse(result, queries)``

File: src/infrastructure/verifyta_verifier.ts [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  ) {
+    this.parser = verifytaOutputParser;
+    this.environment = environment;
+  }
+
+  async verifySolution(
+    xmlFile: string,
+    queries: Array<string>,
+  ): Promise<IQueryResult> {
+    const result = await this.environment.execute(xmlFile);
```
Comment (Brandhoej) [inline src/infrastructure/verifyta_verifier.ts:31]: Ingen fejlhåndtering

File: src/presentation/verifier_controller.spec.ts [hunk 14]
```diff
@@ -0,0 +1,136 @@
+import { describe, expect } from '@jest/globals';
+import fastify from 'fastify';
+import { verifierController } from '.';
+import * as xmlFiles from '../test_files/xml_example_files';
+import * as fs from 'fs';
+import { IQueryResult } from '../infrastructure/i_query_result';
+
+describe('/verifier/verifyta', () => {
```
Comment (Brandhoej) [inline src/presentation/verifier_controller.spec.ts:8]: Der er ingen test som tjekker at resultatet indeholder de korrekte bindings. At skrive ``as xxx`` tjekker ikke for at typen er overholdt for response payload.

File: src/presentation/verifier_controller.ts [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+  // Endpoint to verify uppaal solution using verifyta
+  fastify.post(
+    '/verifiers/:verifier',
+
+    async (request: FastifyRequest, reply: FastifyReply) => {
+      const { verifier } = request.params as { verifier: string };
+
+      // If something other than verifyta is requested
+      if (verifier != 'verifyta') {
+        reply.code(404).send('Verifier not found.');
```
Comment (Brandhoej) [inline src/presentation/verifier_controller.ts:43]: Måske ``Verifier ${verifier} not found``

File: src/presentation/verifier_controller.ts [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+
+      // If no xml in body, reply 400
+      if (xmlInput == null || xmlInput == '') {
+        reply.code(400).send('No xml string in solution field.');
+      }
+
+      const queries = parseQueries(body.queries);
+
+      const result = (
+        await user.verifySolution({
```
Comment (Brandhoej) [inline src/presentation/verifier_controller.ts]: Ingen fejlhåndtering. Det burde given en 500

File: src/presentation/verifier_controller.ts [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+        })
+      ).result;
+
+      // construct response that corresponding to swagger doc
+      const response = {
+        queryResults: Object.fromEntries(result.passedQueriesResults),
+        hasSyntaxError: result.hasSyntaxError,
+        hasParserError: result.hasParserError,
+      };
+      reply.send(response);
```
Comment (Brandhoej) [inline src/presentation/verifier_controller.ts]: Hvor bliver status 200 sat?

File: src/infrastructure/paths.spec.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+});
+
+describe('verifytaLinuxPath', () => {
+  it('Should return the path to the verifyta file', () => {
+    const actual = verifytaLinuxPath();
+    expect(
+      actual.endsWith(`${SLASH}verifyta${SLASH}linux${SLASH}verifyta`),
+    ).toBe(true);
+  });
+});
```
Comment (fatimaosman99) [inline src/infrastructure/paths.spec.ts]: Add handling Darwin (MacOs paths)

File: src/infrastructure/uppaal_xml_model_builder.ts [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  private escape_xml(xml: string): string {
+    xml = xml.replace('&', '&amp;');
+    xml = xml.replace('>', '&gt;');
+    xml = xml.replace('<', '&lt;');
+    xml = xml.replace("'", '&apos;');
+    xml = xml.replace('"', '&quot;');
+    return xml;
+  }
```
Comment (fatimaosman99) [inline src/infrastructure/uppaal_xml_model_builder.ts:23]: Add documentation to this.

File: src/infrastructure/uppaal_xml_model_builder.ts [hunk 20]
```diff
@@ -0,0 +1,103 @@
+export class UppaalXmlModelBuilder {
+  private xml: string;
+  private readonly escap_map: { [mapping: string]: string } = {
+    '&': '&amp;',
+    '>': '&gt;',
+    '<': '&lt;',
+    "'": '&apos;',
+    '"': '&quot;',
+  };
```
Comment (fatimaosman99) [inline src/infrastructure/uppaal_xml_model_builder.ts:9]: This is not used, remove it?

File: src/infrastructure/uppaal_xml_model_builder.ts [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+
+    // If the start idnex is either a "\n" or "\t" then add one to the index
+    const start_char = this.xml.charAt(start_index);
+    if (start_char === '\n' || start_char === '\t') {
+      start_index += 1;
+    } else {
+      start_index -= 1;
+    }
+
+    // Uses clise to "cut away" the "query" tags in "queries"
```
Comment (fatimaosman99) [inline src/infrastructure/uppaal_xml_model_builder.ts:63]:
```suggestion
    // Uses slice to "cut away" the "query" tags in "queries"
```

File: src/infrastructure/uppaal_xml_model_builder.ts [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+    if (!this.has_queries()) {
+      query_tag = `<queries>${query_tag}</queries>`;
+    }
+
+    // The index are of the first character: '<'
+    const end_index = this.has_queries()
+      ? this.xml.indexOf('</queries>')
+      : this.xml.indexOf('</nta>');
+
+    // Uses clise to "cut away" the "query" tags in "queries"
```
Comment (fatimaosman99) [inline src/infrastructure/uppaal_xml_model_builder.ts:93]:
```suggestion
    // Uses slice to "cut away" the "query" tags in "queries"
```

File: src/infrastructure/verifyta_environment.ts [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+   */
+  tempSaveFile(contents: string): string {
+    // Create unique name for the xmlfile
+    const hashedFilename =
+      crypto
+        .createHash('md5')
+        .update(contents + Math.random())
+        .digest('hex') + '.xml';
+
+    // Save temerary file
```
Comment (fatimaosman99) [inline src/infrastructure/verifyta_environment.ts:61]:
```suggestion
    // Save temporary file
```

File: src/infrastructure/paths.ts [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+  if (!process.env.ROOT_PATH) {
+    const srcPath = __dirname.substring(0, __dirname.lastIndexOf(SLASH));
+    return srcPath.substring(0, srcPath.lastIndexOf(SLASH));
+  }
+  return process.env.ROOT_PATH;
+}
+
+export function verifytaLinuxPath(): string {
+  return path.join(rootPath(), 'verifyta', 'linux', 'verifyta');
+}
```
Comment (fatimaosman99) [inline src/infrastructure/paths.ts:16]: Add macOs path

File: src/infrastructure/verifyta_environment.ts [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+   * Runs verifyta on an xmlfile as string. Returns output of verifyta as ICmdResult.
+   * @param xmlFileString The xml file as a string
+   * @returns Promise of cmd result from verifyta
+   */
+  async execute(xmlFileString: string): Promise<ICmdResult> {
+    const filepath = this.tempSaveFile(xmlFileString);
+
+    const verifytaPath =
+      Os.platform() === 'linux' ? verifytaLinuxPath() : verifytaWindowsPath();
+    const command = `${verifytaPath} -u ${filepath}`;
```
Comment (fatimaosman99) [inline src/infrastructure/verifyta_environment.ts:19]: Add the Mac/darwin path when implemented here

**Review by fatimaosman99 — CHANGES_REQUESTED** [review]
Great work. Have few comments :)

File: .eslintrc (hunk 1 above)
Comment (frejaz) [reply .eslintrc:13]: Fordi det er super annoying + det bliver fikset på hvert save

**Pull request closed without merging**